ASAN_TRAP in Style::BuilderConverter::convertInitialLetter

pratiksha-99 opened the pull request:
#### 9649ee33c20c830a78b1fd660123b4a9d2fe5989
<pre>
ASAN_TRAP in Style::BuilderConverter::convertInitialLetter
<a href="https://bugs.webkit.org/show_bug.cgi?id=285459">https://bugs.webkit.org/show_bug.cgi?id=285459</a>
<a href="https://rdar.apple.com/141024836">rdar://141024836</a>

Reviewed by Matthieu Dubet and Ryosuke Niwa.

Added a check to verify the CSSQuad value else throw TypeError.

* LayoutTests/fast/css/css-typed-om/css-builder-converter-initial-letter-crash-expected.txt: Added.
* LayoutTests/fast/css/css-typed-om/css-builder-converter-initial-letter-crash.html: Added.
* Source/WebCore/css/CSSQuadValue.cpp:
(WebCore::CSSQuadValue::CSSQuadValue):
(WebCore::CSSQuadValue::canBeCoalesced const):
* Source/WebCore/css/CSSQuadValue.h:
* Source/WebCore/css/typedom/StylePropertyMap.cpp:
(WebCore::StylePropertyMap::set):

Canonical link: <a href="https://commits.webkit.org/289023@main">https://commits.webkit.org/289023@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f2c6de1ce10e558a6461f0ad2d5176e4744be4c0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/85007 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/4737 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39404 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/90157 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/36063 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/87094 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/4827 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/12713 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/66158 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23976 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/88052 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/3694 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77255 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46423 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/3573 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/31484 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/35136 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/74367 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/32295 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/91579 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/12351 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/9039 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/74680 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/12580 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/73067 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/73797 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/18153 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/16604 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/4368 "The change is no longer eligible for processing.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13260 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/12299 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/17756 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/12135 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/15630 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/13881 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on LayoutTests/fast/dom/StyleSheet/css-builder-converter-initial-letter-crash.html:
```diff
@@ -0,0 +1,24 @@
+<!DOCTYPE html>
```
I think you can just do this instead:
```
<!DOCTYPE html>
<html lang="en">
  <body>
    <p>This test passes if WebKit does not crash.</p>
    <script>
        globalThis.testRunner?.dumpAsText();
        document.body.attributeStyleMap.set('-webkit-initial-letter', document.body.computedStyleMap().get('border-image-outset'));
        document.write('PASS');
    </script>
  </body>
</html>
```

rniwa commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -999,6 +999,9 @@ inline IntSize BuilderConverter::convertInitialLetter(const BuilderState& builde
+
     auto& conversionData = builderState.cssToLengthConversionData();
     return {
         downcast<CSSPrimitiveValue>(value.first()).resolveAsNumber<int>(conversionData),
```
Hm... is it really guaranteed that first & second would be `CSSPrimitiveValue`?

pratiksha-99 replied:
Thats what is being done in `BuilderConverter::convertRadius()` and few other places. We can `dynamicDowncast` but maybe an overkill ?

smfr commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -999,6 +999,9 @@ inline IntSize BuilderConverter::convertInitialLetter(const BuilderState& builde
     if (value.valueID() == CSSValueNormal)
         return IntSize();
 
+    if (!value.isPair())
```
Should this be fixed at the parser level instead?

weinig replied:
Presumably not the parser but in the Typed CSSOM map implementation.

rniwa replied:
You mean that typed OM shouldn't be setting these values to the property?

weinig replied:
Yeah. Or at least, its validation needs to notice this is not valid.

The parser will only ever return a CSSPairValue or CSSValueNormal primitive.

weinig replied:
Is the invalid value it is getting here a CSSQuadValue?

rniwa replied:
Hm... that sounds like a pretty bad design flaw in the current design of Typed OM. We probably need to add a similar validation for all other properties, right?

weinig replied:
I am actually not sure what is going on here, since I think in this case it would actually reparse the value to validate it. Going to try to get it in the debugger and see where the bad value comes from.

weinig replied:
(but, to answer your larger point, the whole premise of Style::BuilderConverter just kind of blindly casting based on knowledge of what the parser / Typed CSSOM might do is bad and really the main goal of my whole strong CSS value representation project)

weinig replied:
Ok, so the code making the invalid assumption is this:

```c++
bool InlineStylePropertyMap::setProperty(CSSPropertyID propertyID, Ref<CSSValue>&& value)
{
    if (!m_element)
        return false;
    StyleAttributeMutationScope mutationScope { m_element.get() };
    bool didFailParsing = false;
    // FIXME: We should be able to validate CSSValues without having to serialize to text and go through the
    // parser. This is inefficient.
    m_element->setInlineStyleProperty(propertyID, value->cssText(), IsImportant::No, &didFailParsing);
    if (!didFailParsing) {
        m_element->setInlineStyleProperty(propertyID, WTFMove(value));
        mutationScope.enqueueMutationRecord();
    }
    return !didFailParsing;
}
```

Specifically, just because setInlineStyleProperty() does not fail parsing, doesn't mean the CSSValue is ok.

Here, what we have is `value` is a CSS `Quad` which serializes to the string "0". So, it parses just fine as a -webkit-initial-letter.

But then passing that `Quad` to `setInlineStyleProperty` is just totally invalid.

So what can we do?

I think the easiest solution here is to do a completely fresh parse into a new CSSValue here. Don't try to re-use it at all.

rniwa replied:
> I think the easiest solution here is to do a completely fresh parse into a new CSSValue here. Don't try to re-use it at all.

What do you mean that? Like check whether the result of setInlineStyleProperty whether it had set the same type??

weinig replied:
It might be as easy as just removing the second call to `setInlineStyleProperty` in that function. I am not sure what it is supposed to be doing. Maybe @cdumez or @anttijk remember.

rniwa replied:
I think it's trying to preserve the identity of `CSSValue`. Otherwise, `getProperty` will return a new `CSSValue`.

mdubet replied:
To get the correct behavior with wrong type (throwing a JS exception) it indeed need to be done earlier in the call stack

mdubet replied:
Oh I haven't seen this conversation sorry. 

Yes we have a frequent issue with the fact that the TypedOM  "type check" is string based, but sometimes the actual type is fairly different from its string representation (pair of numbers coalescing to single number, single number serializing as a pair of similar numbers - which I think is the issue here?) ...etc

The correct behavior is to reject the invalid here types to get the proper JS exception thrown in user land, but it needs an exhaustive mapping of property/type (which doesn't exist currently). 

The way to fix it is to move some of some "tricks" for the CSS parser to the CSS converter (so their work both with CSS stylesheet and CSS OM).

In this case, the trick is to support a single number (being duplicated for the second number) cf the code in the parser for initialLetter

```
RefPtr<CSSValue> consumeWebkitInitialLetter(CSSParserTokenRange& range, const CSSParserContext& context) {
...
    auto height = consumeNumber(range, context, ValueRange::NonNegative);
    if (!height)
        return nullptr;
    RefPtr<CSSPrimitiveValue> position;
    if (!range.atEnd()) {
...
    } else
        position = height.copyRef();
    return CSSValuePair::create(position.releaseNonNull(), height.releaseNonNull());
}
 ````

rniwa replied:
Filed https://bugs.webkit.org/show_bug.cgi?id=285867 to track this design issue.

mdubet commented on LayoutTests/fast/dom/StyleSheet/css-builder-converter-initial-letter-crash.html:
```diff
@@ -0,0 +1,11 @@
+    <p>This test passes if WebKit does not crash.</p>
+    <script>
+        globalThis.testRunner?.dumpAsText();
+        document.body.attributeStyleMap.set('-webkit-initial-letter', document.body.computedStyleMap().get('border-image-outset'));
```
This should throw a JS TypeError exception when the type is incorrect, or setting a correct value.

https://webidl.spec.whatwg.org/#exceptiondef-typeerror

weinig commented on LayoutTests/fast/dom/StyleSheet/css-builder-converter-initial-letter-crash.html:
```diff
@@ -0,0 +1,11 @@
+<!DOCTYPE html>
```
This is kinda a weird place for this test. Should probably be in fast/css/css-typed-om/ (where you can find similar crashes like style-converter-convert-shape-value-type-confusion.html)

mdubet reviewed (changes requested):
Unfortunately, this patch would create a wrong behavior for the Typed CSSOM API. 
The  `attributeMap.set(initial-letter, value)` would silently set a wrong value (0) when given a single number. 

The issue here is that a single number is a valid value for `initial-letter` property (like the more common pair of numbers), so the string based check "is this a correct type for property" pass, but then the code latter doesn't support it.

Supporting a single number here with the correct behavior (which has the behavior of simply duplicating the single number for both numbers in the pair)  would fix this.

https://developer.mozilla.org/en-US/docs/Web/CSS/initial-letter

EDIT : The type here might be a CSS Quad numbers, so the correct fix is to throw a JS TypeError exception.

rniwa commented on Source/WebCore/css/CSSQuadValue.cpp:
```diff
@@ -49,4 +49,15 @@ bool CSSQuadValue::equals(const CSSQuadValue& other) const
 
+bool CSSQuadValue::canBeCoalesced() const
+{
+    if (!m_coalesceIdenticalValues)
```
We never assign a value to this boolean. How does that work??

pratiksha-99 replied:
Whoops, I fixed the constructor which sets the bool. I think it used the default earlier which isn't right.

mdubet commented on Source/WebCore/css/CSSQuadValue.h:
```diff
@@ -37,10 +37,14 @@ class CSSQuadValue final : public CSSValue {
 
+    CSSQuadValue(ValueSeparator, Quad, IdenticalValueSerialization);
+    explicit CSSQuadValue(Quad);
+    bool m_coalesceIdenticalValues { true };
```
Do we ever use the `IdenticalValueSerialization` boolean state and `m_coalesceIdenticalValues` ? It seems we only use the `Coalesce` value, thus always setting the boolean to `true`.

If not, we can just remove them.

pratiksha-99 replied:
We don't. I added to be identical to CSSValuePair. We can always default to true and not set `m_coalesceIdenticalValues` in constructor. I had that code before this commit.

mdubet commented on LayoutTests/fast/css/css-typed-om/css-builder-converter-initial-letter-crash.html:
```diff
@@ -0,0 +1,16 @@
+            if (!(borderImageOutset instanceof CSSStyleValue)) throw new TypeError();
+            document.body.attributeStyleMap.set('-webkit-initial-letter', borderImageOutset);
+        } catch (e) {
+            document.write('PASS: threw TypeError');
```
We don't know what JS exception has been thrown here, it could be something else than the expected `TypeError`.

mdubet replied:
We have `assert_throws_js(TypeError, ....)` to properly assert this.

mdubet replied:
Also, only the `set()` is supposed to throw a `TypeError`, not the `get()`  part

pratiksha-99 replied:
Updated the test, thanks Matthieu.

mdubet commented on LayoutTests/fast/css/css-typed-om/css-builder-converter-initial-letter-crash.html:
```diff
@@ -0,0 +1,16 @@
+      globalThis.testRunner?.dumpAsText();
+        try {
+            const borderImageOutset = document.body.computedStyleMap().get('border-image-outset');
+            if (!(borderImageOutset instanceof CSSStyleValue)) throw new TypeError();
```
We can remove this line

mdubet commented on Source/WebCore/css/CSSQuadValue.cpp:
```diff
@@ -25,18 +25,21 @@
+    , m_coalesceIdenticalValues(serialization != IdenticalValueSerialization::DoNotCoalesce)
     , m_quad(WTFMove(quad))
 {
+    m_valueSeparator = separator;
```
Is it necessary ?

mdubet reviewed (approved):
We should rewrite the test to properly test what we want (that the`set()` throws a `TypeError`).

rniwa commented on LayoutTests/fast/css/css-typed-om/css-builder-converter-initial-letter-crash.html:
```diff
@@ -0,0 +1,19 @@
+      globalThis.testRunner?.dumpAsText();
+      const borderImageOutset = document.body.computedStyleMap().get('border-image-outset');
+      assert_throws_js(TypeError, () => { document.body.attributeStyleMap.set('-webkit-initial-letter', borderImageOutset); }, "attributeStyleMap.set() throws an error");
+      document.write('PASS: threw TypeError');
```
This is rather misleading. This line of code will always execute regardless of whether the exception was thrown or not. We should instead wrap the code in `test` function like this:
```js
test(t => {
  const borderImageOutset = document.body.computedStyleMap().get('border-image-outset');
  assert_throws_js(TypeError, () => {
    document.body.attributeStyleMap.set('-webkit-initial-letter', borderImageOutset);
  }, "attributeStyleMap.set() throws an error");  
});
```
`globalThis.testRunner?.dumpAsText();` and `globalThis.testRunner?.notifyDone();` are unnecessary since `testharnessreport.js` calls them.

pratiksha-99 replied:
Thanks for catching it, I didn't realize while creating this test. Updated.

pratiksha-99 committed:
ASAN_TRAP in Style::BuilderConverter::convertInitialLetter
https://bugs.webkit.org/show_bug.cgi?id=285459
rdar://141024836

Reviewed by Matthieu Dubet and Ryosuke Niwa.

Added a check to verify the CSSQuad value else throw TypeError.

* LayoutTests/fast/css/css-typed-om/css-builder-converter-initial-letter-crash-expected.txt: Added.
* LayoutTests/fast/css/css-typed-om/css-builder-converter-initial-letter-crash.html: Added.
* Source/WebCore/css/CSSQuadValue.cpp:
(WebCore::CSSQuadValue::CSSQuadValue):
(WebCore::CSSQuadValue::canBeCoalesced const):
* Source/WebCore/css/CSSQuadValue.h:
* Source/WebCore/css/typedom/StylePropertyMap.cpp:
(WebCore::StylePropertyMap::set):

Canonical link: https://commits.webkit.org/289023@main

smfr commented on Source/WebCore/css/CSSQuadValue.cpp:
```diff
@@ -25,11 +25,13 @@
 
 CSSQuadValue::CSSQuadValue(Quad quad)
     : CSSValue(ClassType::Quad)
+    , m_coalesceIdenticalValues(true)
```
Why initialize this here, when it has a default initializer in the header?

smfr commented on Source/WebCore/css/CSSQuadValue.h:
```diff
@@ -37,10 +37,11 @@ class CSSQuadValue final : public CSSValue {
 private:
     explicit CSSQuadValue(Quad);
-
+    bool m_coalesceIdenticalValues { true };
```
Is this the best location padding-wise? It might add 7 bytes of padding before the Quad, or it might pack in with a refcount. Worth checking.
